Make MultiSelectEpisodesHelper not a singleton

mchowning opened the pull request:
## Description
This is a follow-up to https://github.com/Automattic/pocket-casts-android/pull/1316. That PR contained quite a few changes, and we found other ways to address the ANRs and the worst multi-select issues, so we didn't move forward with it. We still had quite a few multi-select issues though. In particular, in my testing I observed that multi-select state was getting shared (which would cause things like the toolbar displaying the wrong number of selected items, and other multiselect bugs).

I initially tried porting over a few of the changes from #1316, but I found that they were introducing regressions. It turned out that just simply making the `MultiSelectEpisodesHelper` a singleton solved most of the multi-select issues I see.

This _almost_ fixes https://github.com/Automattic/pocket-casts-android/issues/98, but there is still a multiselect issue with bookmarks that I'm seeing. I'm not trying to address the one remaining multi-select issue in this PR just to keep this PR small and focused.

## Testing Instructions

### 1. Watch for ANR regressions

I think it's very unlikely that this change would cause any ANR issues, but given that we had some ANR issues the last time we made significant changes to multiselect, please keep an eye out for any unexpected slowness as you test this change

### 2. Mult-Select Improvement

I tested this by turning on multiselect in the first screen, navigating to the second screen, and verifying that the second screen was not put into a multiselect state. When running through these tests, make sure you clear multiselect state from the relevant screens before testing (speaking from experience, it's easy to forget to do that and then to think a subsequent test is failing when it really isn't).

Let me know if there are any other test scenarios we should be covering here.

#### 2.1. For each test:
1. Make sure both the start and end screens are not in multi-select mode
2. Enter multiselection on the start screen
3. Navigate to the end screen
4. Verify that the end screen is not in a multi-select state

#### 2.2. Before these changes
1. 🔴 Podcast Bookmarks → Full Screen Player Bookmarks
2. 🟢 Full Screen Player Bookmarks → Podcast Bookmarks
3. 🔴 Podcast Screen → Up Next
5. 🔴 Up Next → Podcast Screen
6. 🔴 Downloads/Starred/Listening-History → Up Next
7. 🔴 Up Next → Downloads/Starred/Listening-History
8. 🔴 Files  → Up Next
9. 🔴 Up Next → Files
10. 🟢 Downloads/Starred/Listening-History → Podcast
11. 🟢 Podcast → Downloads/Starred/Listening-History
12. 🟢 Files → Podcast
13. 🟢 Podcast → Files
14. 🟢 Viewing a podcast on Discover tab → Viewing a podcast on Podcasts tab

#### 2.3. Expected Results After These Changes
1. 🔴 Podcast Bookmarks → Full Screen Player Bookmarks
2. 🟢 Full Screen Player Bookmarks → Podcast Bookmarks
3. 🟢 Podcast Screen → Up Next
4. 🟢 Up Next → Podcast Screen
5. 🟢 Downloads/Starred/Listening-History → Up Next
6. 🟢 Up Next → Downloads/Starred/Listening-History
7. 🟢 Files  → Up Next
8. 🟢 Up Next → Files
9. 🟢 Downloads/Starred/Listening-History → Podcast
10. 🟢 Podcast → Downloads/Starred/Listening-History
11. 🟢 Files → Podcast
14. 🟢 Podcast → Files
15. 🟢 Viewing a podcast on Discover tab → Viewing a podcast on Podcasts tab


## Checklist
- [x] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [X] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [X] I have considered whether it makes sense to add tests for my changes
- [X] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [X] Any jetpack compose components I added or changed are covered by compose previews

pocketcasts commented:
<!--
  0 Errors
  1 Warning: No reviewers have been set for...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">No reviewers have been set for this PR yet. Please request a review from <strong>@ Automattic/pocket-casts-android</strong>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ashiagr reviewed (approved):
This is working well in my testing. 👍 

One issue I noticed:

Earlier `MultiSelectEpisodesHelper` inside `MainActivity`  was used to dismiss the multi-select toolbar from the `UpNext` dialog instead of dismissing the dialog itself on a back gesture. It is redundant now as both do not share multi-select state.

Video from this PR

https://github.com/Automattic/pocket-casts-android/assets/1405144/1247a332-cc98-4f5d-9872-fe7e7d5e7cbc

Video from the app in production

https://github.com/Automattic/pocket-casts-android/assets/1405144/b7e7d495-dcd7-4164-ab0b-bb17674c8acc


I tried to fix this behavior in these commits:
https://github.com/Automattic/pocket-casts-android/pull/1316/commits/f01899298499a3fb87929e7b1178f36f54d8c4ff
https://github.com/Automattic/pocket-casts-android/pull/1316/commits/38b1711a6e9ec82559d48e8c29744fd7ced3402b

It also fixed the back behavior for multi-select toolbar on `Episode Details` and bookmarks from the `Files` bottom sheet. We can cherry-pick the changes here or I can create a new PR if you want.

mchowning commented:
> One issue I noticed:
>
> Earlier MultiSelectEpisodesHelper inside MainActivity was used to dismiss the multi-select toolbar from the UpNext dialog instead of dismissing the dialog itself on a back gesture. It is redundant now as both do not share multi-select state.
> 
> I tried to fix this behavior in these commits...
> https://github.com/Automattic/pocket-casts-android/commit/f01899298499a3fb87929e7b1178f36f54d8c4ff
> https://github.com/Automattic/pocket-casts-android/commit/38b1711a6e9ec82559d48e8c29744fd7ced3402b
>
> It also fixed the back behavior for multi-select toolbar on Episode Details and bookmarks from the Files bottom sheet. We can cherry-pick the changes here or I can create a new PR if you want.

Good catch and those changes look great. I've added them to this PR.

mchowning commented:
You mentioned on Slack that you were interested in adding this to `7.54` since it's going to be a longer code freeze. I've updated the PR to build on `release/7.54` and updated the PR to target `release/7.54`. I'll hold off on merging this just to give you a chance to make sure that this still looks still looks good to you with the changes I've made.

The changes I've made are:
1. Cherry-picked the changes from my PR to be on top of `release/7.54` instead of `main`
2. Cherry-picked your two commits onto this branch
3. Resolved a merge conflict in the `CHANGELOG.md` file
4. Updated the PR to target `release/7.54`
